### PR TITLE
Some modules support no_std

### PR DESCRIPTION
**Pull request by k-nasa**
The following three modules have been changed to work with no_std.

- task
- future
- stream

### CODE REVIEW

**Comment by k-nasa** [comment]
@yoshuawuyts 
I think it's nothing. I think it is enough now.

However, I think it is necessary to add no_std examples and documentation in the future.

Or should it be done now?

File: Cargo.toml [hunk 1]
```diff
@@ -39,13 +39,16 @@ unstable = ["std", "broadcaster", "futures-timer"]
...
-  "futures-core",
   "futures-io",
   "memchr",
   "once_cell",
-  "pin-project-lite",
   "pin-utils",
   "slab",
+  "no-std",
+]
+no-std = [
```
Comment (taiki-e) [inline Cargo.toml]: I believe `alloc` is a better feature name (futures, crossbeam, serde, etc. use this name.).

Comment (k-nasa) [reply Cargo.toml]: That's right. I changed to `alloc`

File: src/utils.rs [hunk 2]
```diff
@@ -1,3 +1,9 @@
+#[cfg(feature = "alloc")]
+extern crate alloc;
```
Comment (taiki-e) [inline src/utils.rs]: Could you move this `extern crate` to `lib.rs` (and remove `#[cfg(feature = "alloc")]`)?

File: src/utils.rs [hunk 3]
```diff
@@ -1,3 +1,9 @@
...
+
+#[cfg(feature = "alloc")]
```
Comment (taiki-e) [inline src/utils.rs]:
```suggestion
```

Explicit import takes precedence over prelude, so this cfg can be removed.

**Comment by k-nasa** [comment]
I couldn't build with target `thumbv7m-none-eabi`.
Looking at build details using `cargo -v`, it appears to be linked `std`.
```
Running `rustc --edition=2018 --crate-name futures_core /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.1/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type lib --emit=dep-info,metadata -C debuginfo=2 --cfg 'feature="alloc"' --cfg 'feature="std"' -C metadata=612df2be1fa85baa -C extra-filename=-612df2be1fa85baa --out-dir /home/runner/work/async-std/async-std/target/thumbv7m-none-eabi/debug/deps --target thumbv7m-none-eabi -L dependency=/home/runner/work/async-std/async-std/target/thumbv7m-none-eabi/debug/deps -L dependency=/home/runner/work/async-std/async-std/target/debug/deps --cap-lints allow -Dwarnings`
```

I ran command  `cargo check --no-default-features --features alloc --target thumbv7m-none-eabi -v`.

output is
```
cargo build --no-default-features --target thumbv7m-none-eabi --features alloc -v
   Compiling futures-core v0.3.1
       Fresh pin-project-lite v0.1.4
     Running `rustc --crate-name futures_core --edition=2018 /Users/nasa/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.1/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C debuginfo=2 --cfg 'feature="alloc"' --cfg 'feature="std"' -C metadata=bf22d8a7eee34ada -C extra-filename=-bf22d8a7eee34ada --out-dir /Users/nasa/lab/oss/k-nasa/async-std/target/thumbv7m-none-eabi/debug/deps --target thumbv7m-none-eabi -L dependency=/Users/nasa/lab/oss/k-nasa/async-std/target/thumbv7m-none-eabi/debug/deps -L dependency=/Users/nasa/lab/oss/k-nasa/async-std/target/debug/deps --cap-lints allow`
error[E0463]: can't find crate for `std`
  |
  = note: the `thumbv7m-none-eabi` target may not be installed

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
error: could not compile `futures-core`.

Caused by:
  process didn't exit successfully: `rustc --crate-name futures_core --edition=2018 /Users/nasa/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.1/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C debuginfo=2 --cfg 'feature="alloc"' --cfg 'feature="std"' -C metadata=bf22d8a7eee34ada -C extra-filename=-bf22d8a7eee34ada --out-dir /Users/nasa/lab/oss/k-nasa/async-std/target/thumbv7m-none-eabi/debug/deps --target thumbv7m-none-eabi -L dependency=/Users/nasa/lab/oss/k-nasa/async-std/target/thumbv7m-none-eabi/debug/deps -L dependency=/Users/nasa/lab/oss/k-nasa/async-std/target/debug/deps --cap-lints allow` (exit code: 1)
```

@taiki-e  Do you know any solution?

**Comment by taiki-e** [comment]
Looks like due to https://github.com/rust-lang/cargo/issues/4866 (cargo leaking *-dependencies into normal build)

Could you add [`-Z avoid-dev-deps`](https://github.com/rust-lang/cargo/issues/5133) flag for cargo command? 

(if we need to run on stable, use [cargo-hack](https://github.com/taiki-e/cargo-hack)'s `--no-dev-deps` flag instead.)

**Comment by k-nasa** [comment]
@taiki-e Thanks!!
I could build no-std.

File: src/future/mod.rs [hunk 4]
```diff
@@ -46,15 +46,20 @@
...
-pub use ready::ready;
+cfg_alloc! {
+    pub use future::Future;
+    pub(crate) mod future;
+}
+
+cfg_std! {
+    pub use pending::pending;
+    pub use poll_fn::poll_fn;
+    pub use ready::ready;
```
Comment (taiki-e) [inline src/future/mod.rs]: FYI: to enable these utilities on no-std, use manually implemented futures instead of async functions.

**Comment by k-nasa** [comment]
Is it a problem to merge this?

**Comment by yoshuawuyts** [comment]
@tirr-c fantastic work; this is great!

**Comment by tirr-c** [comment]
@yoshuawuyts it seems like an autocompletion failure? ;)

**Comment by yoshuawuyts** [comment]
@tirr-c ahaha, oops context switch failure. Was also reviewing Tide patches 😂 sorry sorry!

**Pull request merged**